src/pathfinding/grid_pathfinder.jl

```diff
@@ -0,0 +1,362 @@
+    delta_cost(pathfinder, pathfinder.cost_metric, from, to)
+
+struct GridCell
+    f::Int
+    g::Int
+    h::Int
+end
+
+GridCell(g::Int, h::Int, admissibility::AbstractFloat) =
+    GridCell(round(Int, g + (1 + admissibility) * h), g, h)
+
+GridCell() = GridCell(typemax(Int), typemax(Int), typemax(Int))
+
+"""
+    find_path(pathfinder::AStar{D}, from::NTuple{D,Int}, to::NTuple{D,Int})
+Calculate the shortest path from `from` to `to` using the A* algorithm.
+If a path does not exist between the given positions, an empty linked list is returned.
+
+This function usually does not need to be called explicitly, instead
+the use the provided [`set_target!`](@ref) and [`move_agent!`](@ref) functions.
+"""
+function find_path(pathfinder::AStar{D}, from::Dims{D}, to::Dims{D}) where {D}
+    grid = Dict{Dims{D},GridCell}()
+    parent = Dict{Dims{D},Dims{D}}()
+
+    open_list = MutableBinaryMinHeap{Tuple{Int,Dims{D}}}()
+    open_list_handles = Dict{Dims{D},Int64}()
+    closed_list = Set{Dims{D}}()
+
+    grid[from] = GridCell(0, delta_cost(pathfinder, from, to), pathfinder.admissibility)
+    push!(open_list, (grid[from].f, from))
+
+    while !isempty(open_list)
+        _, cur = pop!(open_list)
+        cur == to && break
+        push!(closed_list, cur)
+
+        nbors = get_neighbors(cur, pathfinder)
+        for nbor in Iterators.filter(n -> inbounds(n, pathfinder, closed_list), nbors)
+            nbor_cell = haskey(grid, nbor) ? grid[nbor] : GridCell()
+            new_g_cost = grid[cur].g + delta_cost(pathfinder, cur, nbor)
+
+            if new_g_cost < nbor_cell.g
+                parent[nbor] = cur
+                grid[nbor] = GridCell(
+                    new_g_cost,
+                    delta_cost(pathfinder, nbor, to),
+                    pathfinder.admissibility,
+                )
+                if haskey(open_list_handles, nbor)
+                    update!(open_list, open_list_handles[nbor], (grid[nbor].f, nbor))
+                else
+                    open_list_handles[nbor] = push!(open_list, (grid[nbor].f, nbor))
+                end
+            end
+        end
+    end
+
+    agent_path = Path{D}()
+    cur = to
+    while true
+        haskey(parent, cur) || break
+        pushfirst!(agent_path, cur)
+        cur = parent[cur]
+    end
+    return agent_path
+end
+
+@inline get_neighbors(cur, pathfinder::AStar{D,true}) where {D} =
+    (mod1.(cur .+ β.I, pathfinder.grid_dims) for β in pathfinder.neighborhood)
+@inline get_neighbors(cur, pathfinder::AStar{D,false}) where {D} =
+    (cur .+ β.I for β in pathfinder.neighborhood)
+@inline inbounds(n, pathfinder, closed) =
+    all(1 .<= n .<= pathfinder.grid_dims) && pathfinder.walkable[n...] && n ∉ closed
+
+"""
+    set_target!(agent::A, target::NTuple{D,Int}, model::ABM{<:GridSpace{D,P,<:AStar{D}},A})
+This calculates and stores the shortest path to move the agent from its current position to `target`
+using [`find_path`](@ref).
+"""
```

Comment:
Okay the docstring of this function needs to be re-written from scratch. First, do it in impreative form " do this ", then, crosslink the `move_along_path!` function instead of the `find_path`. We don't care what internals `set_path!` uses, we care about how a user should use it.

Also say explicitly in the docstring that this function is related with path finding. Don't expect the user to read 3 nested type parameterization, and suddenly see `AStar` and understand everything magically. That won't happen.